SidePanel: Show Stashes

gerhardol opened the pull request:
Fixes #5460 
Follow-up to #10138 (that added stashes to the grid)

## Proposed changes

1. Side panel: Refactor classes and names

Renamed and restructured Node/Tree classes.
Some methods were added on parents but only used in leaves
(especially for submodule vs the "revision grid" branches/tags).
Example for a moved method is aheadBehind for LocalBranchNode.

Renamed and added some classes:
* BaseRevisionTree and BaseBranchNode.cs -> BaseRevisionNode.cs
  for methods pointing to revisions in the grid (i.e. all but submodules).
* BaseRefTree for Refs (Branches/Tags).

Methods were moved to the class where they were used

2. SidePanel:  Visibility for branches and tags was not updated.

If the revision grid had another filter than current branch or
filtered branches, the branches/tags were always set as visible.

The refs are always set as visible at creation and set as
not visible after the grid has loaded.

3.  SidePanel: Show Stashes

Context menu to open/apply/pop/drop as well as stash/stash-staged/manage.

4. RemoteBranches: Show ahead/behind for remotes

Supprt Alt-Click to select the tracking local branches

This is a candidate for 4.0

## Screenshots <!-- Remove this section if PR does not change UI -->

### Before

![image](https://user-images.githubusercontent.com/6248932/195948912-6ed7ea2d-4a92-41ff-a31f-06335669e1a4.png)

Revision filter hiding these branches
![image](https://user-images.githubusercontent.com/6248932/196061745-56b43863-f3a5-44b2-b7a8-d2e30b1eb1f5.png)

### After

![image](https://user-images.githubusercontent.com/6248932/195950064-45ce4cb0-d0ad-430f-97db-c61fdbc33d8a.png)

![image](https://user-images.githubusercontent.com/6248932/196055958-86e4f09e-35f3-48a3-867c-35a2f7ecb101.png)

![image](https://user-images.githubusercontent.com/6248932/196061784-02a9bb86-2faa-4350-b76b-f444b1f9600e.png)

![image](https://user-images.githubusercontent.com/6248932/196813425-669d50de-a49b-4ae8-b660-1abc82f893f1.png)

Remote had no ahaed/behind info (no alt-click)
![image](https://user-images.githubusercontent.com/6248932/197304858-71334c62-ac63-4910-b0fc-2b82101f5731.png)
![image](https://user-images.githubusercontent.com/6248932/197304940-591250ac-c067-47a6-939b-7c47c1bc8dbc.png)

## Test methodology <!-- How did you ensure quality? -->

Tests are adapted to additions.

## Merge strategy

Rebase merge

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

RussKie commented on GitUI/BranchTreePanel/RepoObjectsTree.cs:
```diff
@@ -234,22 +237,39 @@ bool IsOverride(MethodInfo m)
             _tagTree.Refresh(isFiltering, forceRefresh, getRefs);
+
+            // Clear current stash
+            _stashTree.Refresh(new(() => new List<GitRevision>()));
```
Could this be?
```suggestion
            _stashTree.Refresh(new(() => Array.Empty<GitRevision>()));
```

RussKie commented on GitUI/BranchTreePanel/RepoObjectsTree.cs:
```diff
@@ -382,6 +403,18 @@ private void CreateSubmodules()
+                SelectedImageKey = nameof(Images.Stash)
+            };
+
+            _stashTree = new StashTree(rootNode, UICommandsSource, _refsSource);
```
[nit]
```suggestion
            _stashTree = new(rootNode, UICommandsSource, _refsSource);
```

gerhardol replied:
When we talked about this feature, we have previously said that the type should be stated if not on left side of '='
These rules should be documented...
I prefer the suggestion, but do not change right now.

RussKie replied:
We shouldn't use `var` if the type is not obvious from the expression. When newing up the type could be omitted.

This is a non-blocking suggestion.

RussKie commented on GitUI/UserControls/RevisionGrid/RevisionGridControl.cs:
```diff
@@ -887,6 +887,9 @@ public void PerformRefreshRevisions(Func<RefsFilter, IReadOnlyList<IGitRef>> get
             _ambiguousRefs = new(() => GitRef.GetAmbiguousRefNames(getUnfilteredRefs.Value));
 
+            // Get the "main" stash commit, including the reflog selector
+            Lazy<IReadOnlyCollection<GitRevision>> getStashRevs = new(() => (!AppSettings.ShowStashes ? new List<GitRevision>() : new RevisionReader(capturedModule, hasReflogSelector: true).GetStashes(cancellationToken)));
```
```suggestion
            Lazy<IReadOnlyCollection<GitRevision>> getStashRevs =
                new(() => (!AppSettings.ShowStashes 
                    ? Array.Empty<GitRevision>()
                    : new RevisionReader(capturedModule, hasReflogSelector: true).GetStashes(cancellationToken)));
```

I also suspect that one set of parens is superficial.

RussKie commented on GitUI/UserControls/RevisionGrid/RevisionLoadEventArgs.cs:
```diff
@@ -4,12 +4,14 @@ namespace GitUI.UserControls.RevisionGrid
     public class RevisionLoadEventArgs : GitUIEventArgs
     {
-        public RevisionLoadEventArgs(IWin32Window? ownerForm, IGitUICommands gitUICommands, Lazy<IReadOnlyList<IGitRef>> getRefs, bool forceRefresh)
+        public RevisionLoadEventArgs(IWin32Window? ownerForm, IGitUICommands gitUICommands, Lazy<IReadOnlyList<IGitRef>> getRefs, bool forceRefresh, Lazy<IReadOnlyCollection<GitRevision>> getStashRevs)
```
```suggestion
        public RevisionLoadEventArgs(IWin32Window? ownerForm, IGitUICommands gitUICommands, Lazy<IReadOnlyList<IGitRef>> getRefs, Lazy<IReadOnlyCollection<GitRevision>> getStashRevs, bool forceRefresh)
```

RussKie reviewed (approved):
👍 with few nits and suggestions

RussKie commented:
> ![image](https://user-images.githubusercontent.com/6248932/195950154-d119ed77-49e4-4034-b91f-04daf9bd93f5.png)

I don't think it's improving the UI to have the same icon repeated, it's likely achieving the opposite result. We don't have the icons in the toolbar either for the same reason:
![image](https://user-images.githubusercontent.com/4403806/196013365-8db7a5c2-c3ff-4d0c-be47-153b06b5c952.png)

We either need dedicated icons, or have no icons at all.

gerhardol commented on GitUI/BranchTreePanel/StashNode.cs:
```diff
@@ -0,0 +1,70 @@
+    internal sealed class StashNode : BaseBranchNode
+    {
+        public StashNode(Tree tree, in ObjectId? objectId, string reflogSelector, string subject, bool visible)
+            : base(tree, reflogSelector.RemovePrefix("refs/"), visible)
```
#10161 changed to display this in FormStash
![image](https://user-images.githubusercontent.com/6248932/196050239-73b4a397-a249-413e-99d5-35fa1ea34941.png)

The RevGrid show the complete ref (excluding refs/)
![image](https://user-images.githubusercontent.com/6248932/196050299-1cd3167f-d1bf-4568-a285-27db0d29cb9d.png)

Should the list present as in FormStash?
```suggestion
            : base(tree, reflogSelector.RemovePrefix("refs/").RemovePrefix("stash"), visible)
```
![image](https://user-images.githubusercontent.com/6248932/196050334-8795fdce-848a-4c79-aeea-66c4cbfcd9de.png)

mstv replied:
I have no strong opinion regarding the colon - not really needed but for consistency...

```suggestion
            : base(tree, reflogSelector.RemovePrefix(GitRefName.RefsStashPrefix), visible)
```

gerhardol replied:
Changed this, updated screenshot

mstv commented on GitUI/BranchTreePanel/StashTree.cs:
```diff
@@ -0,0 +1,94 @@
+            {
+                token.ThrowIfCancellationRequested();
+
+                // Stashes does not supprt filtering, but stashes may not be visible
```
```suggestion
                // Stashes does not support filtering, but stashes may not be visible
```

mstv commented on GitUI/BranchTreePanel/StashTree.cs:
```diff
@@ -0,0 +1,94 @@
+                // Stashes does not supprt filtering, but stashes may not be visible
+                bool isVisible = stash.ObjectId is not null && _refsSource.Contains(stash.ObjectId);
+                StashNode node = new(this, stash.ObjectId, stash.ReflogSelector, stash.Subject, isVisible);
+                var parent = node.CreateRootNode(pathToNodes, (tree, parentPath) => new BasePathNode(tree, parentPath));
```
```suggestion
                Node? parent = node.CreateRootNode(pathToNodes, (tree, parentPath) => new BasePathNode(tree, parentPath));
```

mstv reviewed (commented):
LGTM, just nits, have not run yet

mstv reviewed (approved):
Thank you, very useful

We should not attempt to select unavailable stash revisions though:

![image](https://user-images.githubusercontent.com/36601201/196058210-ba104043-3f5b-4901-a6b5-ddc1daf9b2ad.png)

gerhardol commented:
> We should not attempt to select unavailable stash revisions though:

You can still "open stash" after dismissing the popup.
The popup should not be raised when right clicking: #10250
This disables double clicking

vbjay commented:
Check button and let through for double click of left / primary button.

gerhardol commented:
Added an incomplete fix for visibility for branches, see updated description.
Normally, the grid is updated just before the branches etc were added. If non-default sorting is used, the sidepanel is updated before the grid and the visibility is incorrect.
For stashes, this was always the case and I added stashes after loading.
The handling was changed for stashes, creating the sidepanel when loading the grid but updating the visibility after the grid was loaded.
Loading earlier is not important but a similar change should be done for branches etc, but there will have to be some more changes.
This second part could be done in a sepate PR too, to get translation changes out.

A related fix:
* Visibility for branches and tags were incorrect if filters other than
branch filers were used.

mstv commented on GitUI/BranchTreePanel/StashTree.cs:
```diff
@@ -63,6 +62,28 @@ private Nodes FillStashTree(IReadOnlyList<GitRevision> stashes, CancellationToke
 
+        internal void UpdateVisibility()
+        {
+            ThreadHelper.JoinableTaskFactory.RunAsync(async () =>
```
❔ `Invoke`

gerhardol replied:
Make this an EventHandler?
What is the benefit in this scenario?

mstv replied:
I think `Invoke` is the short of `RunAsync(await SwitchToMainThread(); actual call;)`. But I don't mind using JTF for all async operations.

gerhardol commented:
It is not obvious how to restructure the info in a good way (to not add null ObjectId to folders, may need to do that anyway or duplicate some code in tags/local/remote/stash).
There is some renaming I would like to do before merging to reflect usage better, will wait a little until I decide...

> Check button and let through for double click of left / primary button.

I guess this was related to https://github.com/gitextensions/gitextensions/pull/10258#pullrequestreview-1143339265, https://github.com/gitextensions/gitextensions/issues/10250
I tried to debug that but did not find the issue.

RussKie commented on GitUI/BranchTreePanel/RepoObjectsTree.cs:
```diff
@@ -234,22 +237,41 @@ bool IsOverride(MethodInfo m)
-        public void Refresh(bool isFiltering, bool forceRefresh, Func<RefsFilter, IReadOnlyList<IGitRef>> getRefs)
+        /// <param name="getRefs">Function to get refs.</param>
+        /// <param name="getStashRevs">Lazy accessor for stash commits.</param>
+        public void RefreshWhenLoading(bool isFiltering, bool forceRefresh, Func<RefsFilter, IReadOnlyList<IGitRef>> getRefs, Lazy<IReadOnlyCollection<GitRevision>> getStashRevs)
```
I find `RefreshWhenLoading` and `RefreshAfterLoaded` confusing. Without reading the xml-docs I couldn't quite figure out what these two do...
I'm having troubles with the meaning of "refreshing while data is loading". `RevisionGrid.RevisionsLoading` event is raised _before_ data is loaded, and `RevisionGrid.RevisionsLoaded` is raised _after_ the data has been loaded and displayed. According to the comment, we have added `RevisionGrid.RevisionsLoading` event specifically to refresh the left panel.

Was there a specific reason why you thought the two methods were necessary? Couldn't we have the single method with an additional flag to indicate whether the data _is being_ or _has been_ loaded?

gerhardol replied:
Before it was Refresh() in many situations which was more confusing.
The first call is before the revision grid is available. The data to add is there, but not if a revision is visible or not, which is available after loading. RefreshAfterLoaded() do not need any parameters right now, it just updates the visibility.

The previous commit updated stash info after loading only. That strategy could be used for all 'commit' (branches, tags, stashes but not submodules), but then updating of will be delayed.

So I prefer two methods, possibly with updated arguments.

gerhardol replied:
Renaming to `RefreshRevisionsLoading` and `RefreshRevisionsLoaded` so similar to the events.
Keeping the arguments the same, as isFiltering is needed after loading (to skip updating visibility).

gerhardol replied:
I renamed the repoObjectsTree method `Refresh` to `RevisionGrid.RevisionsLoading -> RefreshRevisionsLoading` and added `RevisionGrid.RevisionsLoaded -> RefreshRevisionsLoaded`, called from .
The call to RepoObjectTree is now wrapped in `FormBrowse.RefreshLeftPanel`, but keeping the changed names.

RussKie commented on GitUI/UserControls/RevisionGrid/RevisionGridControl.cs:
```diff
@@ -887,6 +887,12 @@ public void PerformRefreshRevisions(Func<RefsFilter, IReadOnlyList<IGitRef>> get
+            // Get the "main" stash commit, including the reflog selector
+            Lazy<IReadOnlyCollection<GitRevision>> getStashRevs = new(() =>
+                !AppSettings.ShowStashes
+                ? new List<GitRevision>()
```
Does it have to be `new List<>`?

gerhardol replied:
Probably not
This is some us for something done once in a process that may take several seconds, so there is not much point in optimizing.
(I have spent a lot of time optimizing also single ms for 4.0, so performance is not unimportant but focus where it makes a difference.)

mdonatas replied:
In this case it's not even about performance. This simply looks wrong. To something which is **ReadOnly** a collection meant for manipulation is assigned. This raises questions and wastes others time while trying to comprehend the meaning of the code.
`Array.Empty<GitRevision>()` set a good tone and communicates intent clearly.

gerhardol replied:
Updating
(I do not find Array.Empty to look better in any way.)

RussKie commented on GitUI/BranchTreePanel/RepoObjectsTree.cs:
```diff
@@ -234,22 +237,41 @@ bool IsOverride(MethodInfo m)
-        public void Refresh(bool isFiltering, bool forceRefresh, Func<RefsFilter, IReadOnlyList<IGitRef>> getRefs)
+        /// <param name="getRefs">Function to get refs.</param>
+        /// <param name="getStashRevs">Lazy accessor for stash commits.</param>
+        public void RefreshWhenLoading(bool isFiltering, bool forceRefresh, Func<RefsFilter, IReadOnlyList<IGitRef>> getRefs, Lazy<IReadOnlyCollection<GitRevision>> getStashRevs)
```
Looks like in different parts we have different order of the same arguments:
![image](https://user-images.githubusercontent.com/4403806/196425121-ac0cba2f-2e14-45cf-a501-d98500d5b9f0.png)

RussKie replied:
> I will not name who set the original order in the event handlers, the side panel methods or who requested changes previously in this PR. We all have a right to change our minds.

😆

gerhardol commented:
Rebased and added three commits:
* review comments
* refactoring without functional changes
* Visibility for branches and tags were incorrect if filters other than branch filers were used.

```
SidePanel: Show Stashes

Context menu to open/apply/pop/drop.

Renamed and restructured Node/Tree classes:
* BaseRevisionTree and BaseBranchNode.cs -> BaseRevisionNode.cs
  for methods pointing to revisions in the grid (i.e. all but submodules).
* BaseRefTree for Refs (Branches/Tags).

Methods were moved to the class where they were used
(like aheadBehind in the LocalBranchNode).

Some renaming to clarify use

* Visibility for branches and tags were incorrect if filters other than
branch filers were used.
```

The Latest refactoring changes some that was added in the first commit and I would like to include it.
I could rebase to make the refactoring first, then 'filter' visibility fixes, then Stash.

There is one omission in features too:

![image](https://user-images.githubusercontent.com/6248932/196554903-0a249e04-7dd8-4bfa-a81e-954d5b03180c.png)
![image](https://user-images.githubusercontent.com/6248932/196554978-3cd96be6-6658-453e-a883-b4cc89c8114c.png)

mstv commented on GitUI/BranchTreePanel/LocalBranchNode.cs:
```diff
@@ -76,5 +81,27 @@ public bool Rename()
+            TreeViewNode.TreeView?.BeginInvoke(new Action(() =>
+            {
+                string branch = RelatedBranch is null || !RepoObjectsTree.ModifierKeys.HasFlag(Keys.Alt)
+                    ? FullPath : RelatedBranch.Substring(startIndex: GitRefName.RefsRemotesPrefix.Length);
```
This used to work for remote branches, too...

gerhardol replied:
Related branch was not set only for Local branch. I was thinking about this myself.
Not tonight, probably separate PR

mstv replied:
OK

gerhardol replied:
I doubt that this has ever worked.
The solution should also eliminate the additional Git command needed for push button vs left panel (as remote/local should share the info anyway). The aheadBehindProvider should be made lazy or the info be provided as the refs. Lazy data reset when the refs are refreshed.

Slightly offtopic, but should remote branches be decorated in some way?
The easiest is to have exactly the same info as locals. The info could be prepended with "local: " or so but that is annoying when you know about it. Tooltip could add details.

mstv commented on GitCommands/Settings/AppSettings.cs:
```diff
@@ -1914,6 +1920,12 @@ public static int RepoObjectsTreeSubmodulesIndex
 
+        public static int RepoObjectsTreeStashesIndex
+        {
+            get => GetInt("RepoObjectsTree.StashesIndex", 4);
```
I have moved it to the top.
(The weird auto-reload of settings without UI update confused me once again. If another GE instance denies to work like for `Move up`, hit `F5`.) 
The `Stashes` tree is collapsed by default. So it does not consume much space but is at hand quickly.
(not requesting a change)

gerhardol replied:
The index is stored in settings, so it is safer to store new occurrences last.
Do you request a change?

mstv replied:
>> (not requesting a change)
>
> Do you request a change?

I just wanted to raise this point. Perhaps others vote for a change.

> The index is stored in settings, so it is safer to store new occurrences last.

Yes, else it might become difficult to handle.

mstv reviewed (commented):
Clear refactoring 👍

I like feature/rot-stashes-2.

gerhardol commented:
Three more commits with one fix, one cleanup in the optimization and addition of context menu on the root stash (screenshot updated).

> Clear refactoring
I like feature/rot-stashes-2.

Squashed and refactored into new commits with refactoring, visibility for branches etc at filtering and stash related separated.
This is how I would like to merge this.

RussKie commented on GitCommands/Git/AheadBehindDataProvider.cs:
```diff
@@ -8,6 +9,7 @@ namespace GitCommands.Git
     public interface IAheadBehindDataProvider
     {
         IDictionary<string, AheadBehindData>? GetData(string branchName = "");
+        void InitLazy();
```
```suggestion
        void Initialize();
```

Whether it's lazy or not is an implementation detail.

gerhardol replied:
ResetCache() or just Reset may be better then - it is resetting the cached result.

RussKie commented on GitCommands/Git/AheadBehindDataProvider.cs:
```diff
@@ -23,20 +25,37 @@ public class AheadBehindDataProvider : IAheadBehindDataProvider
+            // Callers setting branchname has the responisibility to ensure that not all are needed
+            if (string.IsNullOrWhiteSpace(branchName) && !string.IsNullOrWhiteSpace(_branchName))
+            {
+                // Debug.Fail($"Unexpectedly call for all branches after cache filled with specific branch {_branchName}");
```
Is this needed?

gerhardol replied:
The commented out code is to document the behavor. I would have liked to enable the Debug if that had been working in .NET > 4.
For now it auto-resets the cache to get results for the branches.
Should just occur in tests.

RussKie commented on GitUI/BranchTreePanel/BaseBranchLeafNode.cs:
```diff
@@ -51,5 +55,28 @@ public override void ApplyStyle()
+
+        protected override void SelectRevision()
+        {
+            TreeViewNode.TreeView?.BeginInvoke(new Action(() =>
```
It should be possible to write it as:
```suggestion
            TreeViewNode.TreeView?.BeginInvoke(() =>
```

gerhardol replied:
Will change
This was just moved from BaseBranchNode.cs
I assume this goes into the "refactoring - no changes" commit

gerhardol commented:
Pushed some more refactoring - had not planned to do that in this PR, but pushing it here to avoid changing the same class twice.
Some duplicate code i inherited functions could be removed.

The actual change for https://github.com/gitextensions/gitextensions/pull/10258#discussion_r999961425
was then quite small and was included here.

feature/rot-stashes-2 is updated

I still want to include this in 4.0

RussKie commented:
I'd like to get v4 out of the door. We keep dragging on and on with that...

On Sat, 22 Oct 2022 at 11:00, Gerhard Olsson ***@***.***>
wrote:

> Pushed some more refactoring - had not planned to do that in this PR, but
> pushing it here to avoid changing the same class twice.
> Some duplicate code i inherited functions could be removed.
>
> The actual change for #10258 (comment)
> <https://github.com/gitextensions/gitextensions/pull/10258#discussion_r999961425>
> was then quite small and was included here.
>
> feature/rot-stashes-2 is updated
>
> I still want to include this in 4.0
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/10258#issuecomment-1287543773>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABBTEXULK5WA5GBFAX2EYTTWEMVCJANCNFSM6AAAAAARFTNNPU>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

gerhardol commented:
> I'd like to get v4 out of the door. We keep dragging on and on with that...

Agree in general, but this 'completes' a feature

RussKie commented:
OK. Let's get it merged then

On Sat, 22 Oct 2022, 11:32 am Gerhard Olsson, ***@***.***>
wrote:

> I'd like to get v4 out of the door. We keep dragging on and on with that...
>
> Agree in general, but this 'completes' a feature
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/10258#issuecomment-1287557213>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABBTEXUUYXNIASDH6EP2FBDWEMYYTANCNFSM6AAAAAARFTNNPU>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

gerhardol commented:
Need to rebased on master,  #10283
I would like to get a review from @mstv too.
The cache handling in part 4 (that saves a Git call ) can be done in other way, I believe this OK

gerhardol commented:
Squashed to descriptive commits, rebased on master and #10283. Especially #10285 required some redesign.
Some other refining too.

See changes also due to other changes in `feature/rot-stashes-2` that now contain the old PR branch.

gerhardol commented:
Plan to rebase merge this tomorrow, unless there are review comments

mstv commented on GitUI/BranchTreePanel/StashTree.cs:
```diff
@@ -0,0 +1,95 @@
+                return;
+            }
+
+            ThreadHelper.JoinableTaskFactory.RunAsync(async () =>
+            {
+                await ReloadNodesAsync((token, _) =>
+                {
+                    Task<Nodes>? loadNodesTask = null;
+                    _updateSemaphore.WaitAsync();
+                    try
+                    {
+                        loadNodesTask = LoadNodesAsync(token, getStashRevs);
+                    }
+                    finally
+                    {
+                        _updateSemaphore.Release();
+                    }
+
+                    return loadNodesTask;
+                }, null).ConfigureAwait(false);
+            });
```
❔ Do it like in BaseRevisionTree?
```
    <snip>
}).FileAndForget();
```
---

General question: Might `ConfigureAwait()` be the cause why the app sometimes crashes without any popup?

gerhardol replied:
Same for Branches/Tags in BaseRefTree.cs, so not changed.
I guess it could be changed.

> General question: Might ConfigureAwait() be the cause why the app sometimes crashes without any popup?

Maybe Debug.Assert or Debug.Fail that fires if you run in Debug. Hard to debug. Maybe fixed in .net7
https://github.com/dotnet/winforms/issues/3739

mstv reviewed (commented):
Just reviewed on GH

gerhardol commented on GitCommands/Git/AheadBehindDataProvider.cs:
```diff
@@ -23,20 +25,37 @@ public class AheadBehindDataProvider : IAheadBehindDataProvider
+                ResetCache();
+            }
+
+            _lazyData ??= new(() => GetData(null, branchName));
+            _branchName = branchName;
+            return _lazyData.Value;
```
It is a way of syncing.
Something like `_data ??= GetData(null, branchName);` may cause GetData to be called three times, but using Lazy here blocks this.
Could be used with a lock() or so instead.
Can add a comment.

gerhardol commented on GitUI/BranchTreePanel/BaseRefTree.cs:
```diff
@@ -0,0 +1,101 @@
+            });
+        }
+
+        internal override void UpdateVisibility()
+        {
+            if (!IsAttached || !_isFiltering)
+            {
+                return;
+            }
+
+            base.UpdateVisibility();
+        }
```
Reloading is always done before UpdateVisibility(), the visibility for refs is always set to true.

pmiossec commented:
Not a critic but more a remark...

I had to think some longs seconds before understanding why some recent stashes are greyed:
![image](https://user-images.githubusercontent.com/460196/197886517-daea5b1f-5f2f-4442-b446-17fd9b3865f6.png)

before understanding that are the ones based on a deleted commit only reachable when displaying the reflog.

Even if it is right, it is not straightforward and I think we could expect some questions on that after the release 😉 (maybe this comment will be a good answer for those used to search before opening an issue...)

gerhardol commented:
> I had to think some longs seconds before understanding why some recent stashes are greyed: 

There could be a tooltip to enable reflog to see them in the grid, or doubleclick to open the stash viewer (that has improved navigation in 4.0 too).

I assume you expect all to be seen with _All_ branches, branches not visible when branches are filtered have the same decoration.
